runtime/src/chpl-gpu.c

```diff
@@ -389,6 +389,269 @@ void chpl_gpu_comm_get(c_sublocid_t dst_subloc, void *dst,
   }
 }
 
+void chpl_gpu_comm_get_strd(c_sublocid_t dst_subloc,
+                          void* dstaddr_arg, size_t* dststrides,
+                          c_nodeid_t srclocale, c_sublocid_t src_subloc,
+                          void* srcaddr_arg, size_t* srcstrides,
+                          size_t* count, int32_t stridelevels, size_t elemSize,
+                          int32_t commID, int ln, int32_t fn) {
+  // TODO: Re-use code from chpl-comm-strd-xfer.h instead of copying it here.
+  const size_t strlvls=(size_t)stridelevels;
+  size_t i,j,k,t,total,off,x,carry;
+
+  int8_t* dstaddr,*dstaddr1;
+  int8_t* srcaddr,*srcaddr1;
+
+  int *srcdisp, *dstdisp;
+  size_t dststr[strlvls];
+  size_t srcstr[strlvls];
+  size_t cnt[strlvls+1];
+
+
+  // TODO: Communications callback support
+
+  //Only count[0] and strides are measured in number of bytes.
+  cnt[0]=count[0] * elemSize;
+  if(strlvls>0){
+    srcstr[0] = srcstrides[0] * elemSize;
+    dststr[0] = dststrides[0] * elemSize;
+    for (i=1;i<strlvls;i++)
+      {
+        srcstr[i] = srcstrides[i] * elemSize;
+        dststr[i] = dststrides[i] * elemSize;
+        cnt[i]=count[i];
+      }
+    cnt[strlvls]=count[strlvls];
+  }
+
+  switch(strlvls) {
+  case 0:
+    dstaddr=(int8_t*)dstaddr_arg;
+    srcaddr=(int8_t*)srcaddr_arg;
+    chpl_gpu_comm_get(dst_subloc, dstaddr,
+                      srclocale, src_subloc, srcaddr, cnt[0],
+                      commID, ln, fn);
+    break;
+
+  case 1:
+    dstaddr=(int8_t*)dstaddr_arg;
+    srcaddr=(int8_t*)srcaddr_arg;
+    for(i=0; i<cnt[1]; i++) {
+      chpl_gpu_comm_get(dst_subloc, dstaddr,
```

Comment:
I really haven't delved into trying to understand this or the original code in `get_strd_common` but one difference I note is in both versions for case 0 you call `chpl_comm_get` or `chpl_gpu_comm_get` while for all other cases in the original side of things it calls `strd_nb_helper` while in the gpu side of things it calls `chpl_gpu_comm_get`.

So I'm guessing in the original we issue multiple nonblocking communication calls (that we then wait for at the bottom of the function) while for the gpu version we serialize the calls. Was this intentional (maybe because we don't have a good way to do a nonblocking get with GPU's yet?).

If so, I think I'd:

- At the top after your "TODO: Re-use code from chpl-comm-strd-xfer.h instead of copying it" comment expand on that a little explaining how this function differs.  Namely that the gpu version doesn't use nonblocking communication for [some kind of patterns] while non-gpu communication does.
- If you we think this would be a desirable thing for GPUs to do (I'm guessing this impacts performance?) create a GH feature request issue asking for it.

Reply:
Yes, I think you've got it right. The original strided transfer implementation attempts to issue multiple non-blocking calls depending on the comm layer's support via the ``chpl_comm_get/put_nb`` implementations.

I didn't see any non-blocking GPU communication calls and so it seemed easiest to leave that part out. I will add a comment to the top of the functions pointing out this difference.

I'll open an issue about non-blocking GPU calls so that we at least have a place to discuss this.

Reply:
Actually, I might bring this up in a smaller group first to see how feasible that would be, since I don't know much about the GPU runtime implementation.